### PR TITLE
Feature/autoindex extension

### PR DESCRIPTION
**Pull request by boyswan**
This PR provides auto indexing extensions for
- `Catalog` (PartId)
- `Minting` (Id)
- `MultiAsset` (AssetId)

The goal is to align closer to the rmrk-evm spec, and decouple counter logic from the underling implementation.

Each crate provides an `autoindex` extension that can be implemented on a contract, alongside the required AutoIndexData storage. The core implementation now expects Id's to be explicitly passed in, which will allow the user to provide their own or optionally implement the `*AutoIndex` traits to have Ids auto generated (auto increment by 1).

A `Counter` struct has been introduced in `rmrk_common`, which provides logic and is used by all autoindex extensions. 

The `MintingLazy` implementation uses the minting autoindex extension by default. This update also cleans up some duplicate logic with token_id balances, and uses the PSP34 balance data instead of relying on the previous Id storage. I have also removed `mint_many` from the core `Minting` implementation as a user would have to pass in an vector of Id's - this is a user concern and not a lib one,`MintingLazy` still provides `mint_many`.

I have also added an `autoindex` example that shows example methods for `minting` and `multiasset`.

### CODE REVIEW

**Review by Maar-io — COMMENTED** [review]
Can you please elaborate why removing mint_many for eager minting. I find it useful to have it out of box

**Comment by Maar-io** [comment]
In general I see this PR to make the PSP effort more complex

File: crates/multiasset/src/extensions/autoindex.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    types::*,
+};
+
+pub const STORAGE_KEY: u32 = openbrush::storage_unique_key!(MultiAssetAutoIndex);
+
+/// Storage for AutoIndex `AssetId` counter
+#[derive(Default, Debug)]
+#[openbrush::upgradeable_storage(STORAGE_KEY)]
+pub struct MultiAssetAutoIndexData {
+    pub asset_id: Counter<AssetId>,
```
[the remark this answers is not in the document]
Comment (boyswan) [reply crates/multiasset/src/extensions/autoindex.rs:22]: Just to clarify, do you mean comment on the struct, or the `asset_id` field?

**Comment by boyswan** [comment]
> Can you please elaborate why removing mint_many for eager minting. I find it useful to have it out of box

This was mostly to align with the rmrk-evm spec. If it's something we want to keep I can add it back in!

**Review by Maar-io — APPROVED** [review]
LGTM

**Pull request merged**